provisioning/setup.py

```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
+                exit(-1)
+        except Exception as error:
+            logging.error('Error while checking installed rpms (%s)',
+                          error)
+            exit(-1)
+
+# Assumption: motr, hare, consul, cortx-py-utils and cortx-s3server rpms are already installed
+# Approach: Will use 'hctl bootstrap --mkfs <path to CDF>'to start the cluster
+class Init(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            if bootstrap_cluster() != 0:
+                logging.error('Failed to bootstrap the cluster')
+                exit(-1)
+            # if shutdown_cluster() != 0:
+            #     logging.error('Failed to shutdown the cluster')
+            #     exit(-1)
+        except Exception as error:
+            logging.error('Error while initializing the cluster (%s)',
+                          error)
+            exit(-1)
+
+# Assumption: All required rpms are installed and cluster is in running stage
+# Approach: Read CDF file and compare fields with output of 'hctl status'. Check 'check_cluster_status'
+
+class Test(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            if check_cluster_status() != 0:
+                logging.error('Cluster status reports failure')
+                exit(-1)
+        except Exception as error:
+            logging.error('Error while checking cluster status (%s)',
+                          error)
+            exit(-1)
+
+class SupportBundle(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            # Default target directory is /tmp/hare
+            if os.system('hctl reportbug') != 0:
+                logging.error('Failed to generate support bundle')
+                exit(-1)
+        except Exception as error:
+            logging.error('Error while generating support bundle (%s)',
+                          error)
+            exit(-1)
+
+# Approach: Clear hare logs generated until now
+class Cleanup(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            if os.system('rm -f /var/motr/hax/*') != 0:
+                logging.error('Failed to clear Hare logs')
+                exit(-1)
+        except Exception as error:
+            logging.error('Failed to clear Hare logs (%s)',
+                          error)
+            exit(-1)
+
+def checkRpm(rpm_name):
+    return os.system('rpm -qa | grep -q ' + rpm_name)
+
+
+def bootstrap_cluster():
+    return os.system('hctl bootstrap --mkfs /var/lib/hare/cluster.yaml')
+
+
+def shutdown_cluster():
+    return os.system('hctl shutdown')
+
+
+def check_cluster_status():
+    cluster_desc = None
+    with open("/var/lib/hare/cluster.yaml", 'r') as stream:
+        cluster_desc = yaml.safe_load(stream)
+    hostname = socket.gethostname()
```

Comment:
Why do we need to get the current hostname? if we are checking the entire cluster setup the algorithm should be,
- get cluster status `hctl status --json`
- load CDF and iterate over all the nodes,
  - if node runs_confd from CDF, check if corresponding node (using hostname) is running confd and if it is `started` in the dictionary loaded from `hctl status --json` output.
  - check for ioservices for each node.
  
something like,

def check_node_status(node_data_cdf: dict, node_data_hctl: dict) -> bool:
... 

def check_cluster_status() -> bool:
//load cdf
//load `hctl status --json` output
     for node_cdf in cluster_desc['nodes']:
         node_data_cdf = node.get('m0_servers', [])
         node_data_hctl = cluster_info['nodes'][node_data_cdf.hostname]
         if not check_node_status(node_data_cdf, node_data_hctl):
             logging.error()
             return False

Makes sense?

Reply:
@SwapnilGaonkar7 try this,
```
def list2dict(nodes_data_hctl: List) -> dict:
    node_info_dict = {}
    for node in nodes_data_hctl:
        node_svc_info = {}
        for service in node[1]:
            node_svc_info[service[0]] = []
            if (service[3] == 'started'):
                node_svc_info[service[0]].append(service[3])
        node_info_dict[node[0]] = node_svc_info

    return node_info_dict

def check_cluster_status_new() -> bool:
    cluster_desc = None
    with open("/tmp/cluster-6-1-ios.yaml", 'r') as stream:
        cluster_desc = yaml.safe_load(stream)
    cmd = 'hctl status --json'
    cluster_info = json.loads(execute_command_and_get_output(cmd))
    nodes_data_hctl = cluster_info['nodes']

    # node_info_dict = {node[0]: {service[0]: append(service[3]) for service in node[1]} for node in nodes_data_hctl}
    #node_info_dict = {node[0]: {}.setdefault(service[0], []).append(service[3]) for service in node[1] for node in nodes_data_hctl}
    node_info_dict = list2dict(nodes_data_hctl)
    # print(node_info_dict)
    for node in cluster_desc['nodes']:
        m0client_s3_cnt = node['m0_clients']['s3']
        m0client_other_cnt = node['m0_clients']['other']
        m0ds = node.get('m0_servers', [])
        for m0d in m0ds:
            ios_cnt = 0
            if ('runs_confd' in m0d.keys() and
                (node_info_dict[node['hostname']]['confd'][0] != 'started')):
                print('returning false for confd ' + node_info_dict[node['hostname']]['confd'][0])
                return False
            if m0d['io_disks']['data']:
                if node_info_dict[node['hostname']]['ioservice'][ios_cnt] != 'started':
                    print('returning false for ios ' + node['hostname'])
                    return False
                ios_cnt += 1
        if len(node_info_dict[node['hostname']]['s3server']) != m0client_s3_cnt:
            print('returning false for s3servers ' + node['hostname'])
            return False

    return True
```
I specifically omitted m0_clients status check, although its included in the dictionary because they are optional processes.